### PR TITLE
AQI-3744 SDK: Analyte and Test Group Part 2

### PR DESCRIPTION
**Pull request by DarshanaGandhi-AI**

### CODE REVIEW

File: src/Aquarius.ONE.ClientSDK/Operations/Sample/SampleApi.cs [hunk 1]
```diff
@@ -116,18 +115,83 @@ public async Task<bool> CreateTestGroupAsync(TestAnalyteGroup testGroup)
...
+
+        public async Task<bool> UpdateAnalyteAsync(Analyte analyte, DateTime? effectiveDate = null)
+        {
+            var watch = System.Diagnostics.Stopwatch.StartNew();
+
+            var endpoint = $"/operations/sample/v1/analyte/{analyte.Id}";
+
+            if (effectiveDate.HasValue)
+            {
+                endpoint += $"?{effectiveDate.Value.ToString("O")}";
```
Comment (KentMiller) [inline src/Aquarius.ONE.ClientSDK/Operations/Sample/SampleApi.cs:141]: The query parameter should have a name, not just a value. Should be:
$"?effectiveDate={effectiveDate.Value.ToString("O"}";

File: src/Aquarius.ONE.ClientSDK/Operations/Sample/SampleApi.cs [hunk 2]
```diff
@@ -116,18 +115,83 @@ public async Task<bool> CreateTestGroupAsync(TestAnalyteGroup testGroup)
...
+        public async Task<bool> DeleteAnalyteAsync(string analyteId, DateTime? effectiveDate = null)
+        {
+            var watch = System.Diagnostics.Stopwatch.StartNew();
+            var requestId = Guid.NewGuid();
+
+            var endpoint = $"/operations/sample/v1/analyte/{analyteId}";
+
+            if (effectiveDate.HasValue)
+            {
+                endpoint += $"?{effectiveDate.Value.ToString("O")}";
```
Comment (KentMiller) [inline src/Aquarius.ONE.ClientSDK/Operations/Sample/SampleApi.cs:179]: Same comment about including the name and value for the query parameter.

File: src/Aquarius.ONE.ClientSDK/Operations/Sample/SampleApi.cs [hunk 3]
```diff
@@ -207,31 +257,85 @@ public async Task<Analyte> GetOneAnalyteAsync(string analyteId)
...
+
+
+        public async Task<bool> UpdateTestGroupAsync(TestAnalyteGroup testGroup, DateTime? effectiveDate = null)
+        {
+            var watch = System.Diagnostics.Stopwatch.StartNew();
+            var endpoint = $"/operations/sample/v1/testgroup/{testGroup.Id}";
+
+            if (effectiveDate.HasValue)
+            {
+                endpoint += $"?{effectiveDate.Value.ToString("O")}";
```
Comment (KentMiller) [inline src/Aquarius.ONE.ClientSDK/Operations/Sample/SampleApi.cs:284]: Same comment about query parameter name.

File: src/Aquarius.ONE.ClientSDK/Operations/Sample/SampleApi.cs [hunk 4]
```diff
@@ -207,31 +257,85 @@ public async Task<Analyte> GetOneAnalyteAsync(string analyteId)
...
+        public async Task<bool> DeleteTestGroupAsync(string testGroupId, DateTime? effectiveDate = null)
+        {
+            var watch = System.Diagnostics.Stopwatch.StartNew();
+            var requestId = Guid.NewGuid();
+
+            var endpoint = $"/operations/sample/v1/testgroup/{testGroupId}";
+
+            if (effectiveDate.HasValue)
+            {
+                endpoint += $"?{effectiveDate.Value.ToString("O")}";
```
Comment (KentMiller) [inline src/Aquarius.ONE.ClientSDK/Operations/Sample/SampleApi.cs:322]: Same comment

File: src/Aquarius.ONE.ClientSDK/Operations/Sample/SampleCache.cs [hunk 5]
```diff
@@ -272,4 +321,14 @@ private T ErrorResponse<T>(Exception exception, T result)
...
     }
+
+    public class SchedulePropertyBag
+    {
+        public TestGroup TestGroup { get; set; }
+    }
+
+    public class TestGroup
+    {
+        public Guid Id { get; set; }
```
Comment (KentMiller) [inline src/Aquarius.ONE.ClientSDK/Operations/Sample/SampleCache.cs:332]: Is this and the TestGroup in the SchedulePropertyBag class above part of the cache? Should they have public setters?

**Review by KentMiller — CHANGES_REQUESTED** [review]
Please address comments on files and determine if changes are required.

File: src/Aquarius.ONE.ClientSDK/Operations/Sample/SampleApi.cs (hunk 1 above)
Comment (AQI-DanG) [inline src/Aquarius.ONE.ClientSDK/Operations/Sample/SampleApi.cs:141]: Similar to Kent's comments, this is another one where the queryParameter name is not set

File: src/Aquarius.ONE.ClientSDK/Operations/Sample/SampleApi.cs [hunk 6]
```diff
@@ -306,31 +395,49 @@ public async Task<TestAnalyteGroup> GetOneTestGroupAsync(string testGroupId)
...
             }
             catch (Exception e)
             {
-                Event(e, new ClientApiLoggerEventArgs { EventLevel = EnumEventLevel.Error, Module = "SampleApi", Message = $"GetOneTestGroupAsync Failed - {e.Message}" });
+                Event(e, CreateLoggerArgs(EnumEventLevel.Error, $"GetOneTestGroupAsync Failed - {e.Message}"));
+                throw;
+            }
+        }
+
+        public async Task<bool> IsScheduledForUseAsync(string authTwinRefId, string entityType, string entityId)
```
Comment (AQI-DanG) [inline src/Aquarius.ONE.ClientSDK/Operations/Sample/SampleApi.cs:414]: For the SDK would it be better to split this into two methods each with a hardcoded entityType?  Maybe IsAnalyteScheduledForUse and IsTestGroupScheduledForUse?

File: src/Aquarius.ONE.ClientSDK/Operations/Sample/SampleCache.cs [hunk 7]
```diff
@@ -128,6 +130,45 @@ public Activity GetByActivity(string activityId)
             return Activities[Guid.Parse(activityId).ToString()];
         }
 
+        /// <summary>
+        /// Determine if entity is scheduled for use from the cache by entity type (Analyte/TestGroup) and entity Id.
+        /// </summary>
+        /// <param name="entityId">Id of the entity to determine</param>
+        /// <param name="entityType">Entity type: Analyte or TestGroup to determine</param>
+        public bool IsScheduledForUse(string entityType, string entityId)
```
Comment (AQI-DanG) [inline src/Aquarius.ONE.ClientSDK/Operations/Sample/SampleCache.cs:138]: Same comment about potentially splitting this into two methods.

**Review by AQI-DanG — CHANGES_REQUESTED** [review]
To me the purpose of the SDK is to make the apis we have provided even easier for other developers to work with.  By hiding any error codes returned by the apis I think we are doing the opposite of that.  Instead of simply returning false if something fails we should be telling the consumer why it failed, either by passing the error along or via an exception.  I'm not sure it's something that has to change with this story because it a pretty well entrenched pattern in the SDK but I think it is a bad one.  There are still several other code changes that do need to change here to ensure that the correct parameters are being sent.

File: src/Aquarius.ONE.ClientSDK/Operations/Sample/SampleCache.cs (hunk 5 above)
Comment (DarshanaGandhi-AI) [reply src/Aquarius.ONE.ClientSDK/Operations/Sample/SampleCache.cs:332]: If it's private setters then it is not accessible from SampleCache. I want to use it to desearialize the property bag

File: src/Aquarius.ONE.ClientSDK/Operations/Sample/SampleApi.cs [hunk 8]
```diff
@@ -90,12 +91,16 @@ public async Task<bool> CreateAnalyteAsync(Analyte analyte)
...
+                if (!respContent.ResponseMessage.IsSuccessStatusCode)
+                    return ErrorResponse(respContent, false);
+
+                return true;
             }
             catch (Exception e)
             {
                 Event(e, CreateLoggerArgs(EnumEventLevel.Error, $"CreateAnalyteAsync Failed - {e.Message}"));
-                throw;
+                throw e;
```
Comment (KentMiller) [inline src/Aquarius.ONE.ClientSDK/Operations/Sample/SampleApi.cs:103]: Why is this change needed to rethrow the same exception that was caught?

**Review by KentMiller — CHANGES_REQUESTED** [review]
In the exception catch blocks, the change from 'throw' to 'throw e' should be reverted.
Using throw maintains the original stack trace of the exception. When using throw e, the stack original trace is lost.

https://code-maze.com/difference-between-throw-vs-throwex-csharp/

**Comment by AQI-DanG** [comment]
> In the exception catch blocks, the change from 'throw' to 'throw e' should be reverted. Using throw maintains the original stack trace of the exception. When using throw e, the stack original trace is lost.
> 
> https://code-maze.com/difference-between-throw-vs-throwex-csharp/

Thanks Kent, that change was my suggestion when we were brainstorming better ways to report errors to the consumer. I just got those two mixed up.

**Review by AQI-DanG — COMMENTED** [review]
I like the error handling, I think it will be a nice improvement for consumers and gives us a pattern to build on.  As Kent mentioned we probably want to revert the throw changes, sorry for steering you wrong there.

File: src/Aquarius.ONE.ClientSDK/Operations/Sample/SampleApi.cs [hunk 9]
```diff
@@ -116,19 +120,102 @@ public async Task<bool> CreateTestGroupAsync(TestAnalyteGroup testGroup)
...
+                if (!respContent.ResponseMessage.IsSuccessStatusCode)
+                    return ErrorResponse(respContent, false);
+
+                return true;
+
+            }
+            catch (Exception e)
+            {
+                Event(e, CreateLoggerArgs(EnumEventLevel.Error, $"CreateTestGroupAsync Failed - {e.Message}"));
+                throw e;
```
Comment (AQI-DanG) [inline src/Aquarius.ONE.ClientSDK/Operations/Sample/SampleApi.cs:137]: Based on the teams conversation with @JonathanRobertson-AQI  this should also be a call to the ErrorResponse method providing the caught exception and a default return value.

Comment (AQI-DanG) [reply src/Aquarius.ONE.ClientSDK/Operations/Sample/SampleApi.cs:137]: It might be good to have the ErrorResponse method return a custom exception so we do not need to catch things twice.

**Review by KentMiller — APPROVED** [review]

File: src/Aquarius.ONE.ClientSDK/Aquarius.ONE.ClientSDK.csproj [hunk 10]
```diff
@@ -3,9 +3,9 @@
   <PropertyGroup>
     <TargetFramework>netstandard2.0</TargetFramework>
 	  <!--For Release packages update the VersionPrefix and remove any VersionSuffix-->
-	  <VersionPrefix>13.15.1</VersionPrefix>
-	  <!--For Pre-release packages update the VersionSuffix -->
-	  <VersionSuffix></VersionSuffix>
+	  <VersionPrefix>13.16.0</VersionPrefix>
+	  <!--For Feature packages update the VersionSuffix with the Jira ticket number-->
```
Comment (AQI-DanG) [inline src/Aquarius.ONE.ClientSDK/Aquarius.ONE.ClientSDK.csproj:7]: Revert the comment changes please

**Review by AQI-DanG — CHANGES_REQUESTED** [review]
Looks pretty good but you undid my comment change in the csproj so please revert that.

**Review by AQI-DanG — APPROVED** [review]

**Review by JonathanRobertson-AQI — APPROVED** [review]
